misk/src/main/kotlin/misk/scope/ActionScopedProviderModule.kt

```diff
@@ -0,0 +1,112 @@
+        bindSeedData(Key.get(kclass.java, a), Key.get(actionScopedType(kclass), a))
+    }
+
+    /** Binds an annotation qualified [ActionScoped] which only pulls from data seeded at the scope entry */
+    fun <T : Any, A : Annotation> bindSeedData(kclass: KClass<T>, a: KClass<A>) {
+        bindSeedData(Key.get(kclass.java, a.java), Key.get(actionScopedType(kclass), a.java))
+    }
+
+    private fun <T : Any> bindSeedData(key: Key<T>, actionScopedKey: Key<ActionScoped<T>>) {
+        bindProvider(key, actionScopedKey, Provider<ActionScopedProvider<T>> {
+            SeedDataActionScopedProvider(key)
+        })
+    }
+
+    /** Binds an unqualified [ActionScoped] along with its provider */
+    fun <T : Any> bindProvider(
+            kclass: KClass<T>,
+            providerClass: KClass<out ActionScopedProvider<T>>) {
+        bindProvider(
+                Key.get(kclass.java),
+                Key.get(actionScopedType(kclass)),
+                binder().getProvider(providerClass.java))
+    }
+
+    /** Binds an annotation qualified [ActionScoped] along with its provider */
+    fun <T : Any> bindProvider(
+            kclass: KClass<T>,
+            a: Annotation,
+            providerClass: KClass<out ActionScopedProvider<T>>) {
+        bindProvider(
+                Key.get(kclass.java, a),
+                Key.get(actionScopedType(kclass), a),
+                binder().getProvider(providerClass.java))
+    }
+
+    /** Binds an annotation qualified [ActionScoped] along with its provider */
+    fun <T : Any, A : Annotation> bindProvider(
+            kclass: KClass<T>,
+            a: KClass<Annotation>,
+            providerClass: KClass<out ActionScopedProvider<T>>) {
+        bindProvider(
+                Key.get(kclass.java, a.java),
+                Key.get(actionScopedType(kclass), a.java),
+                binder().getProvider(providerClass.java))
+    }
+
+    private fun <T : Any> bindProvider(
+            key: Key<T>,
+            actionScopedKey: Key<ActionScoped<T>>,
+            providerProvider: Provider<out ActionScopedProvider<T>>
+    ) {
+        MapBinder.newMapBinder(binder(), KEY_TYPE, ACTION_SCOPED_PROVIDER_TYPE)
+                .addBinding(key)
+                .toProvider(providerProvider)
+        bind(actionScopedKey).toProvider(RealActionScopedProvider(key)).asSingleton()
+    }
+
+    companion object {
+        @Suppress("UNCHECKED_CAST")
+        private fun <T : Any> actionScopedType(kclass: KClass<T>) =
+                parameterizedType<ActionScoped<T>>(kclass.java).typeLiteral()
+                        as TypeLiteral<ActionScoped<T>>
+
+        private val KEY_TYPE = object : TypeLiteral<Key<*>>() {}
+        private val ACTION_SCOPED_PROVIDER_TYPE = object : TypeLiteral<ActionScopedProvider<*>>() {}
+
+        private class SeedDataActionScopedProvider<T>(private val key: Key<T>) :
+                ActionScopedProvider<T> {
+            override fun get(): T {
```

Comment:
T -> Nothing

Reply:
?

Reply:
`Nothing` is a special feature in Kotlin’s type system. It's a subtype of `T` that is unreachable.